# Exercise 1: MBA

To complete the tasks in this exercise, download the dataset "Market_Basket_Optimisation.csv".

## Task 1 - Answer the question
- What is Market Basket Analysis? Provide an example where this is often used.

Market Basket Analysis (MBA), also known as association rule mining or affinity analysis, is a data mining technique used to discover patterns, relationships, and associations among items or products purchased together in a transaction or shopping basket.

Let's consider a supermarket as an example. In a supermarket, Market Basket Analysis can be applied as follows:

- Discovering Product Associations: By analyzing transaction data, the supermarket may find that customers who purchase bread are likely to buy butter as well (association rule: {Bread} -> {Butter}).
- Cross-Selling: Based on the discovered rules, the supermarket can place bread and butter on nearby shelves or offer discounts when both items are purchased together.

## Task 2 - Analyze the dataset
- Perform market basket analysis on the dataset using the apriori algorithm API from the apyori library package.

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
from apyori import apriori

In [2]:
df = pd.read_csv("Market_Basket_Optimisation.csv", header=None)
df.head(10)

,0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19
0,shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
1,burgers,meatballs,eggs,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,chutney,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,turkey,avocado,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,mineral water,milk,energy bar,whole wheat rice,green tea,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
5,low fat yogurt,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6,whole wheat pasta,french fries,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7,soup,light cream,shallot,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
8,frozen vegetables,spaghetti,green tea,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
9,french fries,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [3]:
df.describe()

,0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19
count,7501,5747,4389,3345,2529,1864,1369,981,654,395,256,154,87,47,25,8,4,4,3,1
unique,115,117,115,114,110,106,102,98,88,80,66,50,43,28,19,8,3,3,3,1
top,mineral water,mineral water,mineral water,mineral water,green tea,french fries,green tea,green tea,green tea,green tea,low fat yogurt,green tea,green tea,green tea,magazines,salmon,frozen smoothie,protein bar,spinach,olive oil
freq,577,484,375,201,153,107,96,67,57,31,22,15,8,4,3,1,2,2,1,1


In [4]:
df.shape

(7501, 20)

In [5]:
df.isnull().sum()

0        0
1     1754
2     3112
3     4156
4     4972
5     5637
6     6132
7     6520
8     6847
9     7106
10    7245
11    7347
12    7414
13    7454
14    7476
15    7493
16    7497
17    7497
18    7498
19    7500
dtype: int64

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7501 entries, 0 to 7500
Data columns (total 20 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       7501 non-null   object
 1   1       5747 non-null   object
 2   2       4389 non-null   object
 3   3       3345 non-null   object
 4   4       2529 non-null   object
 5   5       1864 non-null   object
 6   6       1369 non-null   object
 7   7       981 non-null    object
 8   8       654 non-null    object
 9   9       395 non-null    object
 10  10      256 non-null    object
 11  11      154 non-null    object
 12  12      87 non-null     object
 13  13      47 non-null     object
 14  14      25 non-null     object
 15  15      8 non-null      object
 16  16      4 non-null      object
 17  17      4 non-null      object
 18  18      3 non-null      object
 19  19      1 non-null      object
dtypes: object(20)
memory usage: 1.1+ MB


In [7]:
missing_percent = df.isnull().sum() / len(df) * 100
print(missing_percent)

0      0.000000
1     23.383549
2     41.487802
3     55.405946
4     66.284495
5     75.149980
6     81.749100
7     86.921744
8     91.281163
9     94.734035
10    96.587122
11    97.946940
12    98.840155
13    99.373417
14    99.666711
15    99.893348
16    99.946674
17    99.946674
18    99.960005
19    99.986668
dtype: float64


In [8]:
df_cleaned = df.dropna(axis=1, thresh=len(df) * 0.3)  # Keep columns with at least 30% non-null values

In [9]:
print(df_cleaned.isnull().sum())

0       0
1    1754
2    3112
3    4156
4    4972
dtype: int64


In [10]:
missing_percent = df_cleaned.isnull().sum() / len(df_cleaned) * 100
print(missing_percent)

0     0.000000
1    23.383549
2    41.487802
3    55.405946
4    66.284495
dtype: float64


In [11]:
# For categorical columns (impute with mode)
df_filled = df_cleaned.apply(lambda x: x.fillna(x.mode()[0]) if x.dtype == "object" else x)

In [12]:
df_filled

,0,1,2,3,4
0,shrimp,almonds,avocado,vegetables mix,green grapes
1,burgers,meatballs,eggs,mineral water,green tea
2,chutney,mineral water,mineral water,mineral water,green tea
3,turkey,avocado,mineral water,mineral water,green tea
4,mineral water,milk,energy bar,whole wheat rice,green tea
...,...,...,...,...,...
7496,butter,light mayo,fresh bread,mineral water,green tea
7497,burgers,frozen vegetables,eggs,french fries,magazines
7498,chicken,mineral water,mineral water,mineral water,green tea
7499,escalope,green tea,mineral water,mineral water,green tea


In [13]:
# convert the dataframe into list
transactions = []

for i in range(len(df_filled)):
    transaction = []
    for j in range(5):
        transaction.append(str(df_filled.iloc[i, j]))
    transactions.append(transaction)

In [14]:
transactions[0]

['shrimp', 'almonds', 'avocado', 'vegetables mix', 'green grapes']

In [15]:
# analyze using apriori algorithm

association_rules = apriori(transactions, min_support=0.005, min_confidence=0.3, min_lift=3, min_length=2)
results = list(association_rules)

In [16]:
results

[RelationRecord(items=frozenset({'escalope', 'pasta'}), support=0.005465937874950006, ordered_statistics=[OrderedStatistic(items_base=frozenset({'pasta'}), items_add=frozenset({'escalope'}), confidence=0.37614678899082565, lift=6.297939875491481)]),
 RelationRecord(items=frozenset({'ground beef', 'herb & pepper'}), support=0.015064658045593921, ordered_statistics=[OrderedStatistic(items_base=frozenset({'herb & pepper'}), items_add=frozenset({'ground beef'}), confidence=0.3045822102425876, lift=3.1382845591066615)]),
 RelationRecord(items=frozenset({'tomato sauce', 'ground beef'}), support=0.005199306759098787, ordered_statistics=[OrderedStatistic(items_base=frozenset({'tomato sauce'}), items_add=frozenset({'ground beef'}), confidence=0.37500000000000006, lift=3.8638392857142865)]),
 RelationRecord(items=frozenset({'burgers', 'eggs', 'french fries'}), support=0.005599253432875617, ordered_statistics=[OrderedStatistic(items_base=frozenset({'burgers', 'french fries'}), items_add=frozenset

In [17]:
# Function to extract rules from results
def inspect(results):
    rule_list = []
    for rule in results:
        pair = rule.items  # Base item + associated item
        items = [x for x in pair]
        
        # Rule details
        support = rule.support
        for ordered_stat in rule.ordered_statistics:
            base = list(ordered_stat.items_base)
            add = list(ordered_stat.items_add)
            confidence = ordered_stat.confidence
            lift = ordered_stat.lift
            rule_list.append([base, add, support, confidence, lift])
    
    return pd.DataFrame(rule_list, columns=["Base Item", "Added Item", "Support", "Confidence", "Lift"])

# Convert rules into DataFrame
rules_df = inspect(results)

# Display top 10 rules sorted by lift
rules_df.sort_values(by="Lift", ascending=False).head()


,Base Item,Added Item,Support,Confidence,Lift
0,[pasta],[escalope],0.005466,0.376147,6.297940
2,[tomato sauce],[ground beef],0.005199,0.375000,3.863839
1,[herb & pepper],[ground beef],0.015065,0.304582,3.138285
3,"[burgers, french fries]",[eggs],0.005599,0.456522,3.054745


## Task 3 - Interpret results
- From the results of the analysis done on the dataset, what can you infer from the first result?

**Insights**
- Pasta & Escalope (Lift = 6.30): Customers who buy pasta are highly likely to buy escalope together, suggesting a popular meal pairing.

        ✅ Marketing Strategy: Bundle these items as a meal deal or offer discounts on escalope when purchasing pasta.

- Tomato Sauce & Ground Beef (Lift = 3.86): This combination suggests that customers may be preparing pasta-based meals (e.g., spaghetti bolognese).

        ✅ Marketing Strategy: Place tomato sauce and ground beef near each other in supermarkets or offer recipe-based promotions.

- Herb & Pepper & Ground Beef (Lift = 3.14): This indicates that customers use herb & pepper as a seasoning for meat dishes.

        ✅ Marketing Strategy: Create recipe cards or display seasoning products near the meat section to encourage additional purchases.

### Business Opportunities from These Findings
- **Cross-Selling Strategies**

        - Supermarkets and e-commerce platforms can use smart product placement to boost revenue.
        - Example: Displaying ground beef near tomato sauce can increase sales of both items.

- **Personalized Promotions**

        - Customers who buy pasta frequently can be given discounts on escalope via loyalty programs.
        - Retailers can suggest bundles based on previous purchases.

- **Bundle Offers & Meal Kits**

        - "Pasta Night Kit" → Includes pasta, tomato sauce, and ground beef.
        - "BBQ Kit" → Includes ground beef, herb & pepper seasoning, and burger buns.
        - "Breakfast Special" → Includes burgers, fries, and eggs.